executor/src/witgen/data_structures/caller_data.rs

```diff
@@ -0,0 +1,73 @@
+use itertools::Itertools;
+use powdr_number::FieldElement;
+
+use crate::witgen::{
+    machines::LookupCell,
+    processor::{Left, OuterQuery},
+    EvalError, EvalResult, EvalValue,
+};
+
+/// A representation of the caller's data.
+///
+/// Useful for implementing [Machine::process_plookup] in terms of [Machine::process_lookup_direct].
+pub struct CallerData<'a, 'b, T> {
```

Comment:
Mostly extracted from `call_jit_from_block`, which had these functions [implemented on `OuterQuery`](https://github.com/powdr-labs/powdr/compare/main...call_jit_from_block#diff-b39504252d4fc13559b0d522cf53af70fa9df1a6d616a09c0e5c3a7f5f16d465).

Extracting this exposes the user to fewer details (doesn't need to keep a `Vec<T>`) and decouples the run-time witgen code better from the compile-time witgen code. This struct can be seen as an adapter between the two.